### DISCUSSION TITLE
Lib naming

### DISCUSSION THREAD

**Discussion by Vollstrecker** [pupnp/pupnp#224] [Ideas]
Hi,

I dug some deeper into path-settings of the tests. Current version works if pthread is part of the build, otherwise it fails. I worked on that till it  worked with chainloaded pthread, pthread installed in same prefix and pthread installed elsewhere (there was a bug with finding it, too). All worked in standard Debug build. In Release build the last 2 versions failed. Not with dll-not-found, the told me Entrypoint-not-found.

It took ages to find out, that MS ships an upnp.dll, too. Search order is always Same Dir -> Systems Dir -> Dirs in Path, that why the first version worked, and debug works because of postfix in the name. All I've found regarding this was, that MS claims this one is shipped until Server-2003, nevertheless it's present on my VM with a current Win10. Just VS2019, git and cmake are installed in addition.

To get rid of this problem, the only possible way to prevent this problem would be to name the lib different. I would always prefer to use the projects name, so I would suggest pupnp.

The question is: Would the name-change be more confusing if this would be done for Win only? For linux we could create symlinks from the old name, so every app would work until they are recompiled against the new version. Or we could just let the maintainers do that if needed. If the new version with cmake-config is spread, maintainers should switch to use find_package anyways instead of inventing their own mechanism, so they will have to touch their code this or the other way. For pkg-config based scripts, no change would be needed if it is used as the original plan was. So on linux there wouldn't be a problem. On Win, the prebuild binaries using this lib usually ship the lib with their package, so they can rename it, or adapt the new functionality, for source-distributions applies the same as for linux.

So what's your oppinion on that?

**Comment by mrjimenez** [comment]
Hi Vollstrecker,

Nice research!

From my side, I have no sentimental attachment to the library name, so if it is proven that the best way is to change, I am ok with that.

But as always, lets hear what the other folks think, because it can be a confusing move and we should prepare for it carefully if that is the case.

Regards,
Marcelo.

**Reply by Vollstrecker** [reply]
Sure, the question is, you want additional symlinks, or only namechange on windows?

**Reply by Vollstrecker** [reply]
And for the feedback, gerbera uses it's own find-module that's not searching for the config and not using pkg-config. So this one does the worst-case-scenario for this change.

**Reply by ffontaine** [reply]
gerbera uses pkg-config: https://github.com/gerbera/gerbera/blob/f10c11e8fd6ac9baf6b3d72995e9c6123b7210f1/cmake/Findpupnp.cmake#L10 to retrieve dependencies for static linking.

**Reply by Vollstrecker** [reply]
Nope, they just use pkg-config to get the path's for their own search. And yes, getting what pkg-config returns into a good useable shape isn't that easy.

**Reply by ffontaine** [reply]
Indeed, I missed that, it's strange that we haven't any static build failure on buildroot autobuilders with an openssl-enabled libupnp. I'll double check.

**Reply by ffontaine** [reply]
OK, so we don't have any build failure on buildroot because we're using the latest official release (1.6.4). In this release, the static dependencies (i.e. -lssl, -lcrypto or -latomic) retrieved through pkg-config are used:
https://github.com/gerbera/gerbera/blob/2fa5af758ed840eb35d9e4dbfddb1a2ccacdd798/cmake/Findpupnp.cmake#L38

However, `PC_UPNP_LIBRARIES ` was indeed removed in https://github.com/gerbera/gerbera/commit/628f114e8bd7b087205b37a7285f04b2de684147

So static linking is again broken on upstream gerbera. I'll send a patch or perhaps I should just disable gerbera on buildroot when building statically as it seems that less and less people care about static linking.

**Reply by whyman** [reply]
@ffontaine Hmm, you are right I didnt consider the static linking case when making those changes, I am keen to have it working, will open an issue to remind me to check it.

**Comment by whyman** [comment]
How about calling it `libupnp.dll` on windows too?

**Reply by mrjimenez** [reply]
LOL, that's interesting :smile:

**Reply by Vollstrecker** [reply]
There's nothing that I wouldn't care less, but I think if there is a rename, then we should do it right. The project is called pupnp, so this is the name that just jumps into my mind when I would search for the lib.

I didn't test, but I'm sure, on win cmake searches for the exact given name while in linux lib is prepended, like when it is building and installing, too. So you would have to search for upnp on linux and libupnp on windows. If you have to touch your code, not matter what's the outcome here, I would just change upnp to pupnp in your code (or better add it as alternative name) instead of searching for different names on different plattforms.

For the maintainers, I'll see if someone answers on irc.

**Reply by mrjimenez** [reply]
Sure, I agree with that. We should go to the simpler and less clumsy solution on the long run.

**Comment by mrjimenez** [comment]
My personal preferences are the lowest precedence in this case. I would like that on the long run, we have only one name to link against. If it is necessary to pass through symlinks during some time until distros stabilize, that's ok for me.

Since I am no distro maintainer, I am not at the right position to give pros and cons. I would like to hear from the directly affected people.

**Reply by Vollstrecker** [reply]
Short survey so far, debian doesn't care as long as it's visible in the changelog, suse and fedora don't react at all (on irc general dev of distro, not the maintainers itself) and gentoo it's Ian.

So my proposal would be to just tell them in changelog. Otherwise on the long run we would have deceide how long the links should be shipped and when the right time to change is.

**Reply by mrjimenez** [reply]
If I understood it correctly, the results of this short survey head us to a full rename without links. If that is the case, then I believe we should:

1. Do a last 1.14 release (current master) with the latest bug fixes.
2. Do a 1.16.0 release with the new name.

Is that a good plan?

On my radar I have #218 which is not a release show stopper, since this is a bug that has always existed and could be fixed along 1.16. If anyone has a reason to delay this, please tell me.

**Reply by Vollstrecker** [reply]
Links correct, for the releases: I would prefer to have the changes released before debian code-freeze. They usually miss the first date, but they planned it for 12.02.2021. I mailed the maintainer, as he still is on 1.8.3, but we'll see f he reacts. As their release cycle is so slow, I think it would be better to have this all in there, as dev otherwise would have to maintain old and new options at least until the next debian stable release or until the require some other libs newer as in the yet to be released stable.

**Reply by mrjimenez** [reply]
Well, I don't like the "in a hurry" situation, since it appears to me that these changes should be carefully tested. But we can at least try.

What I can do is release 1.14.1 right now, create a new maintenance branch 1.14.x and start 1.16 in master, what do you think?

**Reply by Vollstrecker** [reply]
To be honest, he didn't react for the notifications about newer versions (or the CVE in 1.12? iirc), so I just mentioned it to have it mentioned. I don't expect him to answer before it's too late anyways. So you're plan is absolutely acceptable.

**Reply by Vollstrecker** [reply]
Little addition: Suse prefers no links, too.

**Reply by mrjimenez** [reply]
1.14.1 is ready, just need to build the tarball. Master (1.16.0) is ready to accept the changes.

**Reply by Vollstrecker** [reply]
I still try to understand the sanitizer build, and maybe there's some more work when the gtest-stuff is ready. So, if it's ok for you, I'll send this all together when it's done.

**Reply by mrjimenez** [reply]
Vollstrecker, please take your time. I just hurried stuff so that you had a chance to have things ready for the Debian release, but I know it is going to be difficult.

**Reply by mrjimenez** [reply]
About the sanitizer build, did you have a look at this: https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html

The sanitizer build we have in github does that:
```bash
./configure --enable-debug \
          CFLAGS="-fsanitize=address,leak" \
          LDFLAGS="-fsanitize=address,leak"
```

**Reply by Vollstrecker** [reply]
That's all? Not that hard.

**Reply by mrjimenez** [reply]
Do notice that you are free to improve it :smiley:

**Reply by Vollstrecker** [reply]
I think you've choosen this by purpose, I just expected way more autotools or libtool voodoo.

**Reply by mrjimenez** [reply]
Lol, that was not even me! :smile:

**Reply by marillat** [reply]
> Links correct, for the releases: I would prefer to have the changes released before debian code-freeze. They usually miss the first date, but they planned it for 12.02.2021. I mailed the maintainer, as he still is on 1.8.3, but we'll see f he reacts. As their release cycle is so slow, I think it would be better to have this all in there, as dev otherwise would have to maintain old and new options at least until the next debian stable release or until the require some other libs newer as in the yet to be released stable.

For Debian it's too late. Library with soname change (transitions) are not allowed since the 2021-01-12
https://release.debian.org/bullseye/freeze_policy.html#transition

**Reply by Vollstrecker** [reply]
Jep, that's why this release happened. To get the fixes and new features in, before a change happens. But I doubt this will happen, as the pkg is in a real bad shape atm.
https://tracker.debian.org/pkg/pupnp-1.8

**Comment by Vollstrecker** [comment]
I've tested now with a virtualbox image provided for testing by MS itself. It worked and didn't have the conflicting file. I'm now going to try to find out where it comes from, but for now: Shame on me, nothing mentioned here is needed atm.
